extensions/exchanges/bitfinex/exchange.js

```diff
@@ -56,201 +248,163 @@ module.exports = function container (get, set, clear) {
+        var args = {}
+        args.sort = -1 //backward
+        args.limit = 1000
+        if (opts.from) {
+          args.start = opts.from
+        }
+        else if (opts.to) {
+          args.end = opts.to
+        }
+        else if (args.start && !args.end) {
+          args.end = args.start + 500000
+        }
+        else if (args.end && !args.start) {
+          args.start = args.end - 500000
+        }
+        var query = encodeQueryData(args)
+        var tpair = 't' + joinProduct(opts.product_id)
+        client.makePublicRequest('trades/' + tpair + '/hist?' + query, function (err, body) {
+          if (err) return retry('getTrades', opts, cb)
+          var trades = body.map(function(trade) {
+            return {
+              trade_id: trade.ID,
+              time: trade.MTS,
+              size: Math.abs(trade.AMOUNT),
+              price: trade.PRICE,
+              side: trade.AMOUNT > 0 ? 'buy' : 'sell'
+            }
+          })
+          cb(null, trades)
+        }) 
+      } else {
+        // We're live now (i.e. opts.from is set), use websockets
+        if (typeof(ws_trades) === "undefined") { return retry('getTrades', opts, cb) }
+        trades = ws_trades.filter(function (trade) { return trade.time >= opts.from })
         cb(null, trades)
-      })
+      }
     },
     
     getBalance: function (opts, cb) {
-      var client = authedClient()
-      client.wallet_balances(function (err, body) {
-        if (err) return(err)
-        var balance = {asset: 0, currency: 0}
-        var accounts = _(body).filter(function (body) { return body.type === c.bitfinex.wallet }).forEach(function (account) {
-          if (account.currency.toUpperCase() === opts.currency) {
-            balance.currency = n(account.amount).format('0.00000000')
-            balance.currency_hold = n(account.amount).subtract(account.available).format('0.00000000')
-          }
-          else if (account.currency.toUpperCase() === opts.asset) {
-            balance.asset = n(account.amount).format('0.00000000')
-            balance.asset_hold = n(account.amount).subtract(account.available).format('0.00000000')
-          }
-        })
-        cb(null, balance)
-      })
+      if (!authed_client_ws) { authedClientWs() }
+      if (Object.keys(ws_balance).length === 0) { return retry('getBalance', opts, cb) }
+
+      balance = {}
+      balance.currency      = n(ws_balance[opts.currency].balance).format('0.00000000')
+      balance.asset         = n(ws_balance[opts.asset].balance).format('0.00000000')
+
+      balance.currency_hold = ws_balance[opts.currency].available ? n(ws_balance[opts.currency].balance).subtract(ws_balance[opts.currency].available).format('0.00000000') : n(0).format('0.00000000')
+      balance.asset_hold    = ws_balance[opts.asset].available    ? n(ws_balance[opts.asset].balance).subtract(ws_balance[opts.asset].available).format('0.00000000')       : n(0).format('0.00000000')
+
+      cb(null, balance)
```

Comment:
Please provide an example, I don't understand the problem.

According to the [bitfinex docs](https://bitfinex.readme.io/v2/reference#ws-auth-wallets), key 4 of the array is 

> Amount not tied up in active orders, positions or funding (null if the value is not fresh enough).

Are we on the same page of the definition of `currency_hold` and `asset_hold`? As far as I understood it, those values are about currency/assets tied up in open/uncanceled/not yet fully returned orders, which we can't use for new orders.

Therefore I think an amount of `0` is correct in most cases.